[SPARK-36811][SQL] Add SQL functions for the BINARY data type for AND, OR, XOR, and NOT

mkaravel opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?

This PR introduces four new SQL functions operating on the BINARY data type.
  1. `bitand`: Takes as input two binary strings and optionally a string argument, and returns their bitwise AND.
  2. `bitor`: Takes as input two binary strings and optionally a string argument, and returns their bitwise OR.
  3. `bitxor`: Takes as input two binary strings and optionally a string argument, and returns their bitwise XOR.
  4. `bitnot`: Takes as input a binary string and returns its bitwise NOT.

In more detail, the first three functions behave as follows:
* If the input is only two binary sequences, they are expected to be of the same byte length, otherwise an exception is thrown. In this case the result is the bitwise operation that the function supports, and the byte length of the resulting byte sequence is equal to the common byte length of the two inputs.
* In the three-argument case, the third argument determines how the operation is performed in the case of byte sequences with unequal lengths. The third argument must be equal to either `lpad` or `rpad` (equality here is case insensitive), otherwise an exception is thrown. More precisely, if the third argument is equal to `lpad`, the shorter byte sequence is conceptually left-padded with zeros to match the length of the longer byte sequence, whereas if the third argument is `rpad`, the shorter byte sequence is conceptually right-padded with zeros to match the length of the longer sequence. Obviously, if the two input byte sequences are of the same length the result is the same for both values of the third argument, and equal to what the two-argument version of the function would have produced. In all cases, the byte length of the result is equal to the maximum of the byte lengths of the two input byte sequences.

#### Examples
* Equal length inputs, two argument overload, `bitand` function:
```scala
scala> sql("select bitand(unhex('aabb'), unhex('03e3'))").show
+--------------------------------------+
|bitand(unhex(aabb), unhex(03e3), lpad)|
+--------------------------------------+
|                               [02 A3]|
+--------------------------------------+
```
* Equal length inputs, three argument overload, `bitand` function:
```scala
scala> sql("select bitand(unhex('aabb'), unhex('03e3'), 'lpad')").show
+--------------------------------------+
|bitand(unhex(aabb), unhex(03e3), lpad)|
+--------------------------------------+
|                               [02 A3]|
+--------------------------------------+

scala> sql("select bitand(unhex('aabb'), unhex('03e3'), 'rpad')").show
+--------------------------------------+
|bitand(unhex(aabb), unhex(03e3), rpad)|
+--------------------------------------+
|                               [02 A3]|
+--------------------------------------+
```
* Unequal length inputs, three argument overload, `bitand` function:
```scala
scala> sql("select bitand(unhex('aabb'), unhex('13e3ff'), 'lpad')").show
+----------------------------------------+
|bitand(unhex(aabb), unhex(13e3ff), lpad)|
+----------------------------------------+
|                              [00 A2 BB]|
+----------------------------------------+

scala> sql("select bitand(unhex('aabb'), unhex('13e3ff'), 'rpad')").show
+----------------------------------------+
|bitand(unhex(aabb), unhex(13e3ff), rpad)|
+----------------------------------------+
|                              [02 A3 00]|
+----------------------------------------+
```
* Unequal length inputs, three argument overload, `bitor` function:
```scala
scala> sql("select bitor(unhex('aabb'), unhex('13e3ff'), 'lpad')").show
+---------------------------------------+
|bitor(unhex(aabb), unhex(13e3ff), lpad)|
+---------------------------------------+
|                             [13 EB FF]|
+---------------------------------------+

scala> sql("select bitor(unhex('aabb'), unhex('13e3ff'), 'rpad')").show
+---------------------------------------+
|bitor(unhex(aabb), unhex(13e3ff), rpad)|
+---------------------------------------+
|                             [BB FB FF]|
+---------------------------------------+
```
* Unequal length inputs, three argument overload, `bitxor` function:
```scala
scala> sql("select bitxor(unhex('aabb'), unhex('13e3ff'), 'lpad')").show
+----------------------------------------+
|bitxor(unhex(aabb), unhex(13e3ff), lpad)|
+----------------------------------------+
|                              [13 49 44]|
+----------------------------------------+


scala> sql("select bitxor(unhex('aabb'), unhex('13e3ff'), 'rpad')").show
+----------------------------------------+
|bitxor(unhex(aabb), unhex(13e3ff), rpad)|
+----------------------------------------+
|                              [B9 58 FF]|
+----------------------------------------+
```

### Why are the changes needed?

These functions are useful for performing bitwise operations on `BINARY` values, seen as bit sets.

Other databases offer similar or the same functionality. In more detail:
* Teradata supports them as [functions](https://docs.teradata.com/r/kmuOwjp1zEYg98JsB8fu_A/zNvtiufFYVtAA~wqxccYKA).
* Postgres supports them as [operators](https://www.postgresql.org/docs/9.4/functions-bitstring.html) over the `BIT` data type with the restriction that the inputs must be of equal length.
* MySQL supports them as [operators](https://dev.mysql.com/doc/refman/8.0/en/bit-functions.html). Arguments must have the same length.
* SQL Server supports them as [operators](https://docs.microsoft.com/en-us/sql/t-sql/language-elements/bitwise-operators-transact-sql?view=sql-server-ver15), but only one of the two arguments can be a binary string.

Oracle, Snowflake, Hive SQL, Big Query do not support the proposed functions for binary strings, but rather only for integral types.

### Does this PR introduce _any_ user-facing change?

Yes. Four new SQL functions.

### How was this patch tested?

Unit tests.

mkaravel committed:
[SPARK-38611][SQL] Add SQL functions for the BINARY data type for AND, OR, XOR, NOT

AmplabJenkins commented:
Can one of the admins verify this patch?

mkaravel committed:
Remove unnecessary formatting comments.

mkaravel committed:
Fix ExpressionSchemaSuite failing test

MaxGekk reviewed (commented):
Any particular reasons to not use the [BitSet](https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html) from JDK?

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:
```diff
@@ -2649,3 +2649,192 @@ case class Sentences(
 }
+
+/**
+ * A function that returns the bitwise AND of two binary strings.
```
The class represents an expression till it is bound to a function name. I would replace `A function ...` -> `The expression ...`

mkaravel replied:
Done. Referring to expressions now.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:
```diff
@@ -2649,3 +2649,192 @@ case class Sentences(
+
+  override def nullSafeEval(left: Any, right: Any): Any = {
+    bytes1.dataType match {
+      case BinaryType => ByteArray.bitwiseAnd(left.asInstanceOf[Array[Byte]],
```
You don't need to check the input type here since you specified the types in `override def inputTypes ...`

mkaravel replied:
Done.

MaxGekk commented:
> Other databases offer similar or the same functionality.

@mkaravel Could you point out which databases have the same, please (with refs to docs).

gatorsmile commented:
add to whitelist

mkaravel committed:
Address review comments:
- Fix code comments to use "expression" instead of "function".
- Remove type checks when not really necessary.

mkaravel commented:
> Any particular reasons to not use the [BitSet](https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html) from JDK?

The way that the `BitSet` class works makes it unsuitable to use here:
* The `and`, `or`, and `xor` in the class produce a Bitset whose size (in bytes) seems to be the minimum of the sizes of the two inputs.
* Trailing zeros in the computed `BitSet` (when using `and`, `or`, or `xor` seem to be discarded after the operation.
* For different sized inputs the bits in `BitSet` are aligned to the left (most significant bit) which does not have the same semantics as the functions implemented in this PR.

For reference, here is the implementation that I used for implementing `bitwiseAnd`:
```java
public static byte[] bitwiseAnd(byte[] bytes1, byte[] bytes2) {
  BitSet bs1 = BitSet.valueOf(bytes1);
  BitSet bs2 = BitSet.valueOf(bytes2);
  bs1.and(bs2);
  return bs1.toByteArray();
}
```

SparkQA commented:
**[Test build #143481 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143481/testReport)** for PR 34056 at commit [`5b8d69f`](https://github.com/apache/spark/commit/5b8d69fac7f43a664ce2e2f74d125d44da8060eb).

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/47992/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/47992/

SparkQA commented:
**[Test build #143481 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143481/testReport)** for PR 34056 at commit [`5b8d69f`](https://github.com/apache/spark/commit/5b8d69fac7f43a664ce2e2f74d125d44da8060eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

mkaravel commented:
> > Other databases offer similar or the same functionality.
> 
> @mkaravel Could you point out which databases have the same, please (with refs to docs).

Updated the PR description with details of what is supported in other systems.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143481/

gatorsmile commented:
cc @cloud-fan

cloud-fan commented on common/unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java:
```diff
@@ -101,4 +101,95 @@ public static long getPrefix(byte[] bytes) {
+    Platform.copyMemory(
+            maxLenBytes, Platform.BYTE_ARRAY_OFFSET,
+            result, Platform.BYTE_ARRAY_OFFSET,
+            maxLen - minLen);
```
is it a no-op if `maxLen - minLen` is 0? or do we need to add a `if-else` manually to avoid calling `copyMemory`?

mkaravel replied:
Yes, it is a no-op. See the code here: https://github.com/apache/spark/blob/d03999ab8846d4897d2ce95ca21a7feed45f292b/common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java#L249

I therefore think there is no need to add an explicit `if-else`. I have also added a few more test cases with equal-length inputs that cover this case.

mkaravel committed:
Added a few more test cases with equal-length inputs.

SparkQA commented:
**[Test build #143657 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143657/testReport)** for PR 34056 at commit [`9339974`](https://github.com/apache/spark/commit/9339974fbfc9effded8c26cbdf47cfb55d197545).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48170/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48170/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48170/

SparkQA commented:
**[Test build #143657 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143657/testReport)** for PR 34056 at commit [`9339974`](https://github.com/apache/spark/commit/9339974fbfc9effded8c26cbdf47cfb55d197545).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143657/

mkaravel commented:
> > Other databases offer similar or the same functionality.
> 
> @mkaravel Could you point out which databases have the same, please (with refs to docs).

Done in the description of the PR.

mkaravel committed:
Changed implementation/API:
- We now have two overloads with 2 and 3 arguments.
- The 2-arg overload accepts only equal length inputs and throws an exception
  otherwise.
- The 3-arg overload accepts a string to indicate how to pad the shorter of
  the two inputs (if they are of unequal length).

mkaravel committed:
Update sql-expression-schema.md

mkaravel committed:
Merge branch 'master' into binary-bitwise-functions

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48903/

mkaravel committed:
Fix failing checkstyle checks.

mkaravel committed:
Fix bugs in Java implementations for BITAND, BITOR, and BITXOR.
Update the unit test results.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48919/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48925/

SparkQA commented:
**[Test build #144446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/144446/testReport)** for PR 34056 at commit [`d945b5e`](https://github.com/apache/spark/commit/d945b5ea549ebde395f74356e625898d1caceba1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144446/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/144452/

mkaravel committed:
Merge branch 'master' into binary-bitwise-functions

mkaravel committed:
Add newline at end of file ByteArray.java

mkaravel committed:
Refactor implementation for BITAND, BITOR, and BITXOR.

mkaravel committed:
Merge branch 'master' into binary-bitwise-functions

mkaravel committed:
Fix Java linter errors.
Updated error messages slightly.

mkaravel committed:
Fix compilation error introduced in previous commit.

mkaravel committed:
Merge branch 'master' into binary-bitwise-functions

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:
```diff
@@ -2734,3 +2734,236 @@ case class Sentences(
+  """,
+  since = "3.3.0",
+  group = "string_funcs")
+case class BitAnd(bytes1: Expression, bytes2: Expression, padding: Expression, numArgs: Int)
```
I think it makes more sense to only allow the padding parameter to be a string constant, so that we can fail earlier (before execution) if the given value is not `lpad` or `rad`.

```
private lazy val isLPad = padding.eval().toString.equalsIgnoreCase("lpad")
private lazy val isRPad = ...
override def checkInputTypes... = {
  ...
  if (!padding.foldable) TypeCheckResult.Fail... else {
    val value = padding.eval()
    if (value == null) TypeCheckResult.Fail... else {
      if (isLPad || isRPad) TypeCheckResult.Success
      else TypeCheckResult.Fail... 
    }
  }
}
```
